### PR TITLE
Add Equip Features to Edit.js

### PR DESCRIPTION
**Pull request by imgntn**
Blocked by this bug: https://highfidelity.fogbugz.com/f/cases/1450/in-vive-hand-controller-clicks-fall-through-html5-overlays-like-edit-js-and-hit-what-s-behind-it

This PR makes it so that you can edit the left & right hand position & rotation of equippable objects in edit.js.  It adds an Equippable toggle and shows/hides fields for the position & rotation for each hand.

To test:
1. Run default scripts from this repository
2. Find an equippable object (ping pong gun, bow and arrow, etc).
3. Open edit and select that item
4. See the new 'Equippable' toggle and section.  If the item is already equippable, that section should be prefilled
5. Make adjustments to the position and rotation of the equip info using the arrows.  Make adjustments using the number input
6. Observe that if you choose a position that would auto-unequip the item, you get an alert and the values revert to their previous values
7. Equip the object and observe that it is in a different place.
8. With an item equipped, make adjustments to the position + rotation.  You should see the object adjust in your hand in realtime.
9. With an item equipped, uncheck 'Equippable'.  The item should unequip.  

https://highfidelity.fogbugz.com/f/cases/1353/Equip-button-in-JS-with-positional-data-field

### CODE REVIEW

**Comment by imgntn** [comment]
A note:  this is not a very satisfying way of making these adjustments.  In the course of building this and discussions with Alan, we propose that we try a new way of making these adjustments:  when a user clicks something in edit, create a left and right hand model overlay in object space that can be adjusted to fit the object.  It makes lots of sense.  That would be a great iteration on this feature.  Even better if the hands can be grabbed and moved in-world.  But its pretty hard to reason about quaternions etc when they are actual numbers.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a685ecf75c9d2be9140c9dca73665e0f0997b320 available at: https://gist.github.com/fb36c33feaea26e1503af6232e0edb3a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8420-a685ecf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8420-a685ecf.dmg

**Comment by Nex-Pro** [comment]
pr #8420

**Testing Architecture:** Desktop PC
**CPU:** 4.4 ghz Intel i7
**RAM:** 16 GB
**GPU:** Nvidia gtx 1070 8gb
**Testing OS:** win 10 pro
**HMD:** HTC Vive.
**Hand Controllers:** Vive
**Leap Motion:** None


### FAIL

Tests:

1. Run default scripts from this repository: **Check**

2. Find an equippable object (ping pong gun, bow and arrow, etc).: **Check**

3. Open edit and select that item: **Check**

4. See the new 'Equippable' toggle and section. If the item is already equippable, that section should be prefilled: **SUCCESS**

5. Make adjustments to the position and rotation of the equip info using the arrows. Make adjustments using the number input: **SUCCESS**

6. Observe that if you choose a position that would auto-unequip the item, you get an alert and the values revert to their previous values: **SUCCESS**

7. Equip the object and observe that it is in a different place.: **FAIL** The object becomes invisible to another user. ( We checked this both ways ) Although it does become offset from original location to the avatar who is holding the object.

8. With an item equipped, make adjustments to the position + rotation. You should see the object adjust in your hand in realtime.: **FAIL** Unable to make any adjustments in the edit panel whilst in HMD as any click on the value or value change arrows results in an object behind the edit panel becoming selected.

9. With an item equipped, uncheck 'Equippable'. The item should unequip.: **SUCCESS**

***We found it impossible to change values in HMD mode.***

***Equipped item became invisible to nearby avatar.***

***One time the item became impossible to detach. Had to close and restart interface.***

**Comment by imgntn** [comment]
thanks @Nex-Pro 
re: 7 that's a strange thing to have happen - so it looks okay to you, but not to someone else?  i'll investigate on monday
re: 8 that's a separate bug with its own ticket about edit.js having clicks fall through.  already logged, it's a bad one!

**Comment by Nex-Pro** [comment]
@imgntn Yep I was testing it with MichelleLeckrone we both saw the pop-gun vanish after adjusting the Y offset. When that occured I didn't test for X or Z. 
As for 8 yep it's bad.

As for the design of this my feeling is that we should have two methods of adjustment. 

One is to toggle an Equip-Adjust mode where the user can re-position the item with the other hand. I feel that would be fairly intuitive. And probably fairly easy to implement.

Second one with a move widget that can be toggled to a rotate widget. This would probably use the maya or blender style of widget ( I think there is an outstanding worklist job for that )

Oh and thanks for working on this it is something that is really going to help.

**Comment by Nex-Pro** [comment]
@imgntn My reasoning for using two adjust methods is that some people will only have the use of one hand.  On thinking further about this it might also be an idea to be able to say use the grip to trigger a freeze in position of the equipped item then simply move the hand to the required position. So there are several approaches that can work.

**Comment by imgntn** [comment]
Actually your first method is already there -- if you have something equipped in your hand you can use your other hand to adjust it.  That's why there's code about protecting from an "auto-unequip" -- if you're adjusting it in your hand and it moves too far out, it unequips.  So that's already there (yay)
 
The second method is similar to what I just built but with some kind of slider / other UI instead of the number inputs.  That would definitely be an improvement but still involves a lot of fiddling with knobs.

There's two parts to this:  1) Equip Points 2)Position/Rotation in hands

re:1) alan and i have been thinking about ways to perhaps do away with this, we'll share soon.  re:2) I think you'll like the 'equip hands' idea once we prototype it -- essentially a visual representation of exactly where the position/rotation will be when you pick it up.  Assuming the visual hands are as easy to rotate/move as a normal object it should work well.

**Comment by Nex-Pro** [comment]
@imgntn Nice, why didn't I try that?

**Comment by Nex-Pro** [comment]
@imgntn Ok I tried the adjust with other hand and It is pretty much spot on. Thanks for that.  I've only had a Vive for a week or so so still not fully up to speed on all the nuances.

**Comment by hifi-gustavo** [comment]
Build console output for commit: fcec89078f2130c5d247685e208dae06a1bec7b9 available at: https://gist.github.com/fb36c33feaea26e1503af6232e0edb3a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8420-fcec890.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8420-fcec890.dmg

**Comment by imgntn** [comment]
i tested here using two machines and didn't see anything weird with the equipping, myself.  can we do another QA pass on this?

**Comment by hifi-gustavo** [comment]
Build console output for commit: bf1eb55ca259f24929b7511fad265b823c7dfc5e available at: https://gist.github.com/fb36c33feaea26e1503af6232e0edb3a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8420-bf1eb55.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8420-bf1eb55.dmg

**Comment by Nex-Pro** [comment]
@imgntn With regard to that vanishing of equipped items that was Both ways. If Michelle was wearing it I could not see it and vice versa. We saw this with the ping-pong gun in the default content apartment.

**Comment by Nex-Pro** [comment]
@imgntn  We can confirm that the vanishing of equipped items is not down to this PR we also saw this in 5277 and seems to be in some way linked to the orientation relative to the hand of the equippable item. We found we could make items vanish by simply equipping upside down or at 90 degrees.

However this is a **FAIL* due to the impossibility of changing any values in the entity properties window of the equipped item whilst wearing an HMD. We both attempted this multiple times and were continuously frustrated when all that occured was the selection of a background object through the properties window. Are we missing something ? Is there some toggle ( I  couldn't find one ) that sets the selected item to remain so?

**Comment by imgntn** [comment]
@Nex-Pro thanks for the thorough QA.  the clicks falling through bug is known and serious.  it's assigned to someone so hopefully it will get fixed soon.  thanks! https://highfidelity.fogbugz.com/f/cases/1450/in-vive-hand-controller-clicks-fall-through-html5-overlays-like-edit-js-and-hit-what-s-behind-it

**Comment by imgntn** [comment]
@Nex-Pro holding the item upside down in a certain way and pressing equip should release an item when its equipped, but definitely shouldnt disappear it if you're picking it up for the first time.  def a bug

File: scripts/system/html/entityProperties.html [hunk 1]
```diff
@@ -74,20 +70,33 @@
             <label for="property-hyperlink-href">Href - hifi://address</label>
             <input type="text" id="property-hyperlink-href">
         </div>
-
+        <div class="hyperlink-group hyperlink-section property text">
+            <label for="property-hyperlink-description">Description</label>
+            <input type="text" id="property-hyperlink-description">
+        </div>
```
Comment (thoys) [inline scripts/system/html/entityProperties.html]: Yo James, moving the description field back down? ;)

**Comment by imgntn** [comment]
hmm yeah i guess that snuck back in.  i don't think that this branch is likely to get merged; adjusting equip stuff this way is just too awkward.

**Pull request closed without merging**